plugins/kotlinx-serialization/testData/boxIr/delegatedProperty.kt

```diff
@@ -0,0 +1,86 @@
+// TARGET_BACKEND: JVM_IR
+
+// WITH_STDLIB
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.json.*
+import kotlin.reflect.KProperty
+import kotlin.properties.*
+
+@Serializable
+data class SimpleDTO(
+    val realProp: Int,
+) {
+    @Transient
+    private val additionalProperties: Map<String, Int> = mapOf("delegatedProp" to 123)
+    @Transient
+    val delegatedProp: Int? by additionalProperties
+}
+
+// optimized properties must also work
+// https://kotlinlang.org/docs/whatsnew1720.html#more-optimized-cases-of-delegated-properties
+// A named object:
+object NamedObject {
+    operator fun getValue(thisRef: Any?, property: KProperty<*>): String = "test-string"
+}
+
+@Serializable
+data class DelegatedByObjectProperty(
+    val realProp: Int
+) {
+    val delegatedProp: String by NamedObject
+}
+
+// A final val property with a backing field and a default getter in the same module:
+val impl: ReadOnlyProperty<Any?, String> = object : ReadOnlyProperty<Any?, String> {
```

Comment:
I see you've added more test-cases for optimized properties, which is good. However, I also suggested that you make sure that even non-optimized delegated properties are still working correctly. Fortunately, it is easy to do so: just change `val` here to `var` and synthetic `delegatedProp$delegate` will appear in `DelegatedByFinalVal` - without breaking the test. (don't forget to rename the class itself though, as it will be no longer ByFinalVal)

Reply:
Not quite sure what do you mean here. You want me to change existing test to use `var` or add another test like `DelegateByVar` that does same thing but with top level var?

Reply:
> add another test like DelegateByVar that does same thing but with top level var?

I think that would be great

Reply:
Ok

Reply:
Before committing/pushing anything can I ask you to verify that I've understood you correctly? Here is what I'm about to add to `delegatedProperty.kt` file:
```kotlin
// A var property with a backing field and a default getter in the same module:
val implvar: ReadWriteProperty<Any?, String> = object : ReadWriteProperty<Any?, String> {
    private var value = "test-string"
    override operator fun getValue(thisRef: Any?, property: KProperty<*>): String = value
    override operator fun setValue(
        thisRef: Any?,
        property: KProperty<*>,
        value: String) { this.value = value }
}

@Serializable
class DelegatedByVar(
    val realProp: Int
) {
    val delegatedProp: String by implvar
}
...

    val byVar = DelegatedByVar(123)
    val byVarJsonStr = Json.encodeToString(byVar)
    val byVarDecoded = Json.decodeFromString<DelegatedByObjectProperty>(byVarJsonStr)
    if (byVarJsonStr != """{"realProp":123}""") return simpleDTOJsonStr
    if (byVarDecoded.delegatedProp != byVar.delegatedProp) return "DelegatedByVar Delegate is incorrect!"
    if (byVarDecoded.realProp !== 123) return "DelegatedByVar Deserialization failed"
```
I don't quite understand what this should do though... Why delegate being a var makes a difference (won't its value just get baked in into DelegatedByVar instance upon construction?)

Reply:
> Why delegate being a var makes a difference

See this section: https://kotlinlang.org/docs/whatsnew1720.html#more-optimized-cases-of-delegated-properties

`val` delegate is being optimized, while `var` is not. You can see this with `println(DelegatedByVar::class.java.declaredFields.toList())`: `ByVal` will not have `delegatedProp$delegate` field, but `ByVar` will have. You also can examine bytecode of both classes.

Reply:
I see, though your answer is not exactly answers "why it makes a difference" :) I assume it will not be optimized (which means it still will be generating hidden `$delegate` field) because it must store current value of delegate var at runtime (when `DelegatedByVar` get constructed) while for `val` case this is guaranteed to be immutable and therefore instead of storing current value we can always use whatever was in the val definition and therefore hidden field is redundant.

Reply:
Serialization plugin generates a special constructor, which does not delegate to default and therefore should correctly initialize all fields (including synthetic once, like `$delegate`). Frankly speaking, I thought that test with non-optimized property would fail because you skip all `it.isDelegate()` properties and therefore $delegate field may be uninitialized, but it seems some other mechanism is doing this, therefore your change is correct.